### PR TITLE
mir: Improve size_of handling when arg is unsized

### PR DESCRIPTION
**Pull request by osa1**
As discussed on Zulip with @RalfJung.

### CODE REVIEW

**Comment by osa1** [comment]
@RalfJung this seems to make the error worse though. Should I maybe try to find out why we try to evaluate this stuff even though the program has type errors? (problem described in https://rust-lang.zulipchat.com/#narrow/stream/269128-miri/topic/Help.20fixing.20.2380742/near/223243016 IIUC)

File: compiler/rustc_mir/src/interpret/step.rs [hunk 1]
```diff
@@ -266,10 +266,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 let layout = self.layout_of(ty)?;
                 if layout.is_unsized() {
                     // FIXME: This should be a span_bug (#80742)
-                    self.tcx.sess.delay_span_bug(
-                        self.frame().current_span(),
-                        &format!("SizeOf nullary MIR operator called for unsized type {}", ty),
-                    );
+                    throw_inval!(SizeOfUnsizedType(ty));
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/step.rs:269]: Please keep the `delay_span_bug`, and then throw the error after that. (That's what we do for `transmute`.)

**Comment by RalfJung** [comment]
> this seems to make the error worse though

You mean because there's two errors now? Hm. Still that seems better than `SizeOf` just making up a nonsensical return value.^^

File: compiler/rustc_mir/src/interpret/step.rs (hunk 1 above)
Comment (osa1) [reply compiler/rustc_mir/src/interpret/step.rs:273]: Done.

**Comment by osa1** [comment]
> You mean because there's two errors now? Hm. Still that seems better than SizeOf just making up a nonsensical return value.^^

Yeah. I started looking into this but it's going slow as I don't have much experience with the code base. If anyone have any tips that would be helpful.

**Comment by RalfJung** [comment]
Thanks!
@bors r+

**Pull request merged**